llvm_passes/HipPromoteInts.cpp

```diff
@@ -0,0 +1,309 @@
+            // Create operation in promoted type
+            NewInst = Builder.CreateBinOp(BinOp->getOpcode(), LHS, RHS);
+        } else {
+            // For operations that should stay in original type
+            if (LHS->getType() != BinOp->getType())
+                LHS = Builder.CreateTrunc(LHS, BinOp->getType());
+            if (RHS->getType() != BinOp->getType())
+                RHS = Builder.CreateTrunc(RHS, BinOp->getType());
+            NewInst = Builder.CreateBinOp(BinOp->getOpcode(), LHS, RHS);
+        }
+        
+        errs() << Indent << "  " << *I << "    ============> " << *NewInst << "\n";
+        PromotedValues[I] = NewInst;
+        Replacements.push_back(Replacement(I, NewInst));
+    }
+    else if (isa<CallInst>(I)) {
+        CallInst* OldCall = cast<CallInst>(I);
+        // Create a new call with the same operands, but use promoted values where available
+        SmallVector<Value*, 8> NewArgs;
+        for (unsigned i = 0; i < OldCall->arg_size(); ++i) {
+            Value* OldArg = OldCall->getArgOperand(i);
+            Value* NewArg = PromotedValues.count(OldArg) ? PromotedValues[OldArg] : OldArg;
+            
+            // if the function expects a non-standard type, abort for now.
+            // TODO: if this assert is hit, we need to handle this case in the future
+            // by promoting the function arguments as well.
+            if (OldArg->getType() != NewArg->getType())
+                assert(false && "HipPromoteIntsPass: Function expects non-standard type");
+                
+            NewArgs.push_back(NewArg);
+        }
+        
+        CallInst* NewCall = CallInst::Create(OldCall->getFunctionType(),
+                                           OldCall->getCalledOperand(),
+                                           NewArgs,
+                                           OldCall->getName(),
+                                           OldCall);
+        NewCall->setCallingConv(OldCall->getCallingConv());
+        NewCall->setAttributes(OldCall->getAttributes());
+        
+        
+        errs() << Indent << "  " << *I << "    ============> " << *NewCall << "\n";
+        PromotedValues[I] = NewCall;
+        Replacements.push_back(Replacement(I, NewCall));
+    }
+    else {
+        assert(false && "HipPromoteIntsPass: Unhandled instruction type");
+    }
+}
+
+bool promoteChain(Instruction *OldI, Type *NonStdType, Type *PromotedTy,
+                      SmallPtrSetImpl<Instruction*> &Visited,
+                      SmallVectorImpl<Replacement> &Replacements,
+                      SmallDenseMap<Value*, Value*> &PromotedValues,
+                      unsigned Depth = 0) {
+    // If we've already processed this instruction, just return
+    if (!Visited.insert(OldI).second) {
+        // If we have a promoted value for this instruction, use it
+        if (PromotedValues.count(OldI)) {
+            errs() << std::string(Depth * 2, ' ') << "Already processed: " << *OldI << "\n";
```

Comment:
This produces spam so perhaps remove it or warp it in LLVM_DEBUG. `dbgs()` is more appropriate for printing debug messages.

Reply:
These are temporary - btw, how do you get  LLVM_DEBUG. dbgs() to print?

Reply:
By passing `-debug` to `opt` or `llc` tool, or `-mllvm -debug` to `clang` tool.